### PR TITLE
Mt-KaHyPar Strong Integration

### PR DESCRIPTION
**Pull request by kittobi1992**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,54 +33,57 @@ undone and, at each level, several *local search* method are used to improve the
...
-instances that could not be partitioned within the time limit are shown with a clock symbol.
-
-To compare us against different sequential hypergraph partitioner, we use a benchmark set consisting of 488 hypergraphs (see [Benchmark Statistics][SetA]). In the figures, we compare Mt-KaHyPar with the sequential hypergraph partitioners
-PaToH 3.3 in quality (PaToH-Q), default (PaToH-D) and speed mode (PaToH-S), the recursive bisection variant (hMETIS-R) of hMETIS 2.0 and
-KaHyPar-CA (similiar algorithmic components as Mt-KaHyPar) and KaHyPar-HFC (extends KaHyPar-CA with flow-based refinement) of the
+We use the [*performance profiles*](https://link.springer.com/article/10.1007/s101070100263) to compare Mt-KaHyPar Fast and Strong to other partitioning algorithms in terms of solution quality (for a detailed explanation see either linked paper or our publications).
```
Comment (larsgottesbueren) [inline README.md:61]: superfluous 'the'

File: README.md [hunk 2]
```diff
@@ -33,54 +33,57 @@ undone and, at each level, several *local search* method are used to improve the
...
+
+To compare us against different sequential hypergraph partitioner, we use a benchmark set consisting of 488 hypergraphs (see [Benchmark Statistics][SetA], refered to as set A). In the figures, we compare Mt-KaHyPar Fast and Strong with the sequential hypergraph partitioners
```
Comment (larsgottesbueren) [inline README.md:63]: to compare us against --> to compare with

File: README.md [hunk 3]
```diff
@@ -33,54 +33,57 @@ undone and, at each level, several *local search* method are used to improve the
...
+with 94 large hypergraph (see [Benchmark Statistics][SetB], refered to as set B), we compare us against the distributed hypergraph partitioner Zoltan 3.83 and the default preset of PaToH 3.3.
+
+Comparing all sequential hypergraph partitioners with Mt-KaHyPar Fast and Strong (using 10 threads) on set A:
+
+<img src="https://user-images.githubusercontent.com/9654047/105867822-cb74fc80-5ff5-11eb-8ae1-9bf92257ab9f.png" alt="alt text" width="100%" height="100%">
+
+Mt-KaHyPar Strong produces partitions with comparable quality to the sequential high-quality hypergraph partitioners hMetis-R and KaHyPar-CA, while being a factor of 30 resp. 8 faster on average (see running time plot below). Mt-KaHyPar Fast produces partitions with comparable quality to the quality preset of PaToH with a running time comparable to the default preset of PaToH. Mt-KaHyPar Strong computes significantly better partitions than Mt-KaHyPar Fast. The sequential high-quality hypergraph partitioner KaHyPar-HFC still computes better partitions than Mt-KaHyPar Fast and Strong. However, KaHyPar-HFC is on average 15 times
```
Comment (larsgottesbueren) [inline README.md:73]: can we reduce the information here? also we're somewhat better than PaToH-Q

File: README.md [hunk 4]
```diff
@@ -98,26 +101,40 @@ The following command will install most of the required dependencies:
...
+Building Mt-KaHyPar
 -----------
 
 1. Clone the repository including submodules:
 
    ```git clone --depth=1 --recursive git@github.com:kittobi1992/mt-kahypar.git```
 2. Create a build directory: `mkdir build && cd build`
 3. Run cmake: `cmake .. -DCMAKE_BUILD_TYPE=RELEASE`
-4. Run make: `make KaHyPar`
+4. Run make: `make MtKaHyPar -j8`
```
Comment (larsgottesbueren) [inline README.md:112]: why tell people how many threads to use? I think make -j gives you max parallelism of the machine.

File: README.md [hunk 5]
```diff
@@ -98,26 +101,40 @@ The following command will install most of the required dependencies:
...
+
+The build produces two executables, which will be located in `build/mt-kahypar/application/`:
 
-The binary will be located in `build/mt-kahypar/application/`. Per default, MT-KaHyPar uses 32-bit vertex and hyperedge IDs. If you want to partition hypergraphs with more than 4.294.967.295 vertices or hyperedges, add option `-DKAHYPAR_USE_64_BIT_IDS=ON` to the `cmake` build command.
+- `MtKaHyParFast`: A scalable hypergraph partitioner that computes good partitions very fast
+- `MtKaHyParStrong`: A scalable hypergraph partitioner that computes high-quality partitions
 
-Running MT-KaHyPar
+Per default, Mt-KaHyPar uses 32-bit vertex and hyperedge IDs. If you want to partition hypergraphs with more than 4.294.967.295 vertices or hyperedges, add option `-DKAHYPAR_USE_64_BIT_IDS=ON` to the `cmake` build command.
```
Comment (larsgottesbueren) [inline README.md:119]: this information can go at the end. tell people how to get it running first

File: README.md [hunk 6]
```diff
@@ -98,26 +101,40 @@ The following command will install most of the required dependencies:
...
+Mt-KaHyPar has several configuration parameters. We recommend to use our presets which are placed in the `config` folder:
+
+- `fast_preset.ini`: Contains the default parameters for Mt-KaHyPar Fast (`MtKaHyParFast`)
+- `strong_preset.ini`: Contains the default parameters for Mt-KaHyPar Strong (`MtKaHyParStrong`)
 
-To run Mt-KaHyPar, you can use the following command:
+If you want to change parameters manually, please run `./MtKaHyParFast --help` or `./MtKaHyParStrong --help` for a detailed description of the different program options. We use the [hMetis format](http://glaros.dtc.umn.edu/gkhome/fetch/sw/hmetis/manual.pdf) for the input hypergraph file as well as the partition output file.
```
Comment (larsgottesbueren) [inline README.md:129]: this is good. multiple students have asked how to find out about the options

File: README.md [hunk 7]
```diff
@@ -228,6 +245,8 @@ To remove the library from your system use the provided uninstall target:
 make uninstall-kahypar
 ```
 
+Note, our library interfaces uses Mt-KaHyPar Fast. We are currently working on a solution to also integrate Mt-KaHyPar Strong.
```
Comment (larsgottesbueren) [inline README.md:248]: why not two library targets? also we might want to let the binaries just use the corresponding library --> compile only once. I think KaHiP does this.

File: README.md [hunk 8]
```diff
@@ -239,13 +258,26 @@ Licensing
...
+      pages     = {16--30},
+      year      = {2021},
+      publisher = {SIAM}
+    }
+
 Contributing
 ------------
 If you are interested in contributing to the Mt-KaHyPar framework
-feel free to contact me or create an issue on the
+feel free to contact us or create an issue on the
```
Comment (larsgottesbueren) [inline README.md:280]: Hehe, nice

File: tests/end_to_end/mt_kahypar_integration_tests.cc [hunk 9]
```diff
@@ -124,52 +127,7 @@ using MultiLevelKm1Config = TestConfig<k,
...
-                                             true,
-                                             InitialPartitioningMode::recursive_bisection>,
-                         MultiLevelCutConfig<4,
-                                             true,
-                                             InitialPartitioningMode::recursive_bisection>,
-                         MultiLevelCutConfig<8,
-                                             true,
-                                             InitialPartitioningMode::recursive_bisection>,
-                         MultiLevelKm1Config<2,
+typedef ::testing::Types<MultiLevelKm1Config<2,
```
Comment (larsgottesbueren) [inline tests/end_to_end/mt_kahypar_integration_tests.cc:130]: Good catch! In the improved_gain_recalculation there should be a commit that keeps these tests but deactivates FM refinement if cut objective is set. Might want to cherry pick that.

File: README.md (hunk 3 above)
Comment (kittobi1992) [reply README.md:73]: yes but this is not easy to see from plots

**Pull request merged**